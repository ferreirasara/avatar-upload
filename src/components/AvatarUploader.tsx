import '../i18n/config';
import { createContext, useCallback, useContext, useState } from "react";
import { useDropzone } from 'react-dropzone';
import { ImageContainer } from './ImageContainer';
import { AvatarUploaderComponentState, AvatarUploaderContainerProps, TAvatarUploaderContext } from '../@types/AvatarUploader';
import { useTranslation } from 'react-i18next';
import { ReactComponent as ImageFileIcon } from '../icons/image-file.svg';
import { ReactComponent as CloseIcon } from '../icons/close.svg';
import { Slider } from './Slider';

const AvatarUploaderContext = createContext<TAvatarUploaderContext>({});

const AvatarUploaderContainer = (props: AvatarUploaderContainerProps) => {
  const { children, bordered } = props;
  return <div className={'avatar-uploader-container' + (bordered ? ' bordered' : '')}>
    {children}
  </div>
}

const CloseButton = () => {
  const { handleCancel } = useContext(AvatarUploaderContext);

  return <div className='close-button-container'>
    <CloseIcon onClick={handleCancel} />
  </div>
}

const AvatarUploaderInner = () => {
  const { componentState } = useContext(AvatarUploaderContext);

  switch (componentState) {
    case 'initial':
      return <AvatarUploaderInitial />
    case 'error':
      return <AvatarUploaderError />
    case 'cropping':
      return <AvatarUploaderCrop />

    default:
      return <AvatarUploaderInitial />;
  }
}

const AvatarUploaderError = () => {
  const { t } = useTranslation();

  return <>
    <ImageContainer />
    <div className='error-text-container'>
      <span className='error-text'>{t('errorState.sorry')}</span>
      <br />
      <span className="try-again-text">{t('errorState.tryAgain')}</span>
    </div>
    <CloseButton />
  </>
}

const AvatarUploaderCrop = () => {
  const { uploadedImage } = useContext(AvatarUploaderContext);
  const { t } = useTranslation();
  const [zoomValue, setZoomValue] = useState<number>(50);



  return <>
    <ImageContainer image={uploadedImage} />
    <div className='crop-container'>
      <span className="crop-text">
        {t('croppingState.crop')}
      </span>
      <Slider value={zoomValue} onChange={(newValue) => setZoomValue(newValue)} />
      <div className='save-crop-button-container'>
        <button className="save-crop-button">
          {t('croppingState.save')}
        </button>
      </div>
    </div>
    <CloseButton />
  </>
}

const AvatarUploaderInitial = () => {
  const { uploadedImage, handleUploadeImage } = useContext(AvatarUploaderContext);
  const { t } = useTranslation();

  const onDrop = useCallback(<T extends File>(acceptedFiles: T[]) => {
    acceptedFiles?.forEach(file => handleUploadeImage && handleUploadeImage(file));
  }, [handleUploadeImage])

  const { getRootProps, getInputProps } = useDropzone({
    onDrop,
    accept: { 'image/jpeg': ['.jpeg', '.png'] }
  })

  return <div
    {...getRootProps()}
    className='dropzone-container'
  >
    <input {...getInputProps()} />
    {uploadedImage ? <ImageContainer image={uploadedImage} /> : null}
    <div className='dropzone-text-container'>
      <span className='initial-text primary'>
        <ImageFileIcon /> {t('initialState.logo')}
      </span>
      <br />
      <span className='initial-text secondary'>
        {t('initialState.dropImage')}
      </span>
    </div>
  </div>
}

export const AvatarUploader = () => {
  const [componentState, setComponentState] = useState<AvatarUploaderComponentState>('initial');
  const [uploadedImage, setUploadedImage] = useState<string>('');

  const handleCancel = () => setComponentState('initial');

  const handleUploadeImage = (file: File) => {
    setUploadedImage(URL.createObjectURL(file));
    setComponentState('cropping');
  };

  return <AvatarUploaderContainer bordered={componentState === 'initial'}>
    <AvatarUploaderContext.Provider value={{ componentState, uploadedImage, handleCancel, handleUploadeImage }}>
      <AvatarUploaderInner />
    </AvatarUploaderContext.Provider>
  </AvatarUploaderContainer>
}