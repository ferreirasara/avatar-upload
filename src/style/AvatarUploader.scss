body {
  font-family: Inter;
}

.dropzone-container {
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  gap: 32px;
}

.dropzone-text-container {
  flex-grow: 1;
}

.initial-text {
  font-size: 14px;
  line-height: 25.2px;

  &.primary {
    color: #495567;
    font-weight: 500;
  }

  &.secondary {
    color: #677489;
    font-weight: 400;
  }
}

.error-text-container {
  text-align: justify;
  flex-grow: 1;
}

.error-text {
  font-family: 'Inter';
  font-style: normal;
  font-weight: 400;
  font-size: 16px;
  line-height: 180%;
  letter-spacing: -0.02em;
  color: #C64D32;
}

.try-again-text {
  font-family: 'Inter';
  font-style: normal;
  font-weight: 500;
  font-size: 16px;
  line-height: 180%;
  letter-spacing: -0.02em;
  text-decoration-line: underline;
  color: #3D485F;
}

.crop-container {
  display: flex;
  flex-grow: 1;
  height: 100%;
  flex-direction: column;
  justify-content: space-between;
  text-align: justify;
}

.crop-text {
  font-weight: 400;
  font-size: 16px;
  line-height: 28.8px;
  color: #677489;
  letter-spacing: -0.02em;
}

.save-crop-button-container {
  display: flex;
  width: 100%;
  justify-content: flex-end;
}

.save-crop-button {
  font-family: 'Inter';
  appearance: none;
  background: #3D485F;
  border-radius: 16px;
  padding: 2px 36px;
  font-family: 'Inter';
  font-style: normal;
  font-weight: 500;
  font-size: 14px;
  line-height: 180%;
  letter-spacing: -0.02em;
  color: #FFFFFF;
  border: none;
}

.close-button-container {
  display: flex;
  height: 100%;
}