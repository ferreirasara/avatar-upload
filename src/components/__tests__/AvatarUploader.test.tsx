import renderer from 'react-test-renderer';
import { AvatarUploader } from '../AvatarUploader';

describe('AvatarUploader', () => {
  it('renders correctly', () => {
    const avatarUploader = renderer.create(<AvatarUploader />).toJSON();
    expect(avatarUploader).toMatchSnapshot();
  });
})

