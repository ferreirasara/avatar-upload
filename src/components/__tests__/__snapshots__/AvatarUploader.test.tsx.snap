// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarUploader renders correctly 1`] = `
<div
  className="avatar-uploader-container bordered"
>
  <div
    className="dropzone-container"
    onBlur={[Function]}
    onClick={[Function]}
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    role="presentation"
    tabIndex={0}
  >
    <input
      accept="image/jpeg,.jpeg,.png"
      multiple={true}
      onChange={[Function]}
      onClick={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      tabIndex={-1}
      type="file"
    />
    <div
      className="dropzone-text-container"
    >
      <span
        className="initial-text primary"
      >
        <svg>
          image-file.svg
        </svg>
         
        Organization Logo
      </span>
      <br />
      <span
        className="initial-text secondary"
      >
        Drop the image here or click to browse.
      </span>
    </div>
  </div>
</div>
`;
